src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/PatchResourceHandler.cs

```diff
@@ -0,0 +1,109 @@
+using Microsoft.Health.Fhir.Core.Models;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Health.Fhir.Core.Features.Resources.Patch
+{
+    public class PatchResourceHandler : BaseResourceHandler, IRequestHandler<PatchResourceRequest, PatchResourceResponse>
+    {
+        private readonly IModelInfoProvider _modelInfoProvider;
+        private readonly ResourceDeserializer _resourceDeserializer;
+        private readonly FhirJsonParser _fhirJsonParser;
+
+        public PatchResourceHandler(
+            IFhirDataStore fhirDataStore,
+            Lazy<IConformanceProvider> conformanceProvider,
+            IResourceWrapperFactory resourceWrapperFactory,
+            ResourceIdProvider resourceIdProvider,
+            IAuthorizationService<DataActions> authorizationService,
+            IModelInfoProvider modelInfoProvider,
+            FhirJsonParser fhirJsonParser,
+            ResourceDeserializer resourceDeserializer)
+            : base(fhirDataStore, conformanceProvider, resourceWrapperFactory, resourceIdProvider, authorizationService)
+        {
+            EnsureArg.IsNotNull(modelInfoProvider, nameof(modelInfoProvider));
+            EnsureArg.IsNotNull(fhirJsonParser, nameof(fhirJsonParser));
+
+            _modelInfoProvider = modelInfoProvider;
+            _resourceDeserializer = resourceDeserializer;
+            _fhirJsonParser = fhirJsonParser;
+        }
+
+        public async Task<PatchResourceResponse> Handle(PatchResourceRequest message, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(message, nameof(message));
+
+            if (await AuthorizationService.CheckAccess(DataActions.Write, cancellationToken) != DataActions.Write)
+            {
+                throw new UnauthorizedFhirActionException();
+            }
+
+            ResourceKey key = message.ResourceKey;
+
+            if (!string.IsNullOrEmpty(key.VersionId))
+            {
+                throw new MethodNotAllowedException(Core.Resources.DeleteVersionNotAllowed);
+            }
+
+            ResourceWrapper currentDoc = await FhirDataStore.GetAsync(key, cancellationToken);
+
+            if (currentDoc == null)
+            {
+                throw new ResourceNotFoundException(string.Format(Core.Resources.ResourceNotFoundById, key.ResourceType, key.Id));
+            }
+
+            if (currentDoc.IsHistory)
+            {
+                throw new MethodNotAllowedException(Core.Resources.PatchVersionNotAllowed);
+            }
+
+            try
+            {
+                FhirJsonNode node = (FhirJsonNode)FhirJsonNode.Parse(currentDoc.RawResource.Data);
+                JObject nodeJson = node.ToJObject();
+                string versionId = (string)nodeJson["meta"]["versionId"];
+                string resourceId = message.ResourceKey.Id;
+
+                message.PatchDocument.ApplyTo(nodeJson);
+
+                FhirJsonNode nodePatch = (FhirJsonNode)FhirJsonNode.Create(nodeJson);
```

Comment:
nit: this one can just be
```suggestion
                ISourceNode jsonStatement = FhirJsonNode.Create(nodeJson);
```